torchgen/dest/dispatchless.py

```diff
@@ -0,0 +1,265 @@
+            and not f.func.is_out_fn()
+        )
+        return index.has_kernel(g) or cond_meta or cond_composite
+    return index.has_kernel(f)
+
+
+@dataclass(frozen=True)
+class DispatchlessComposite:
+    # dispatch key we are generating for.
+    dispatch_key: DispatchKey
+
+    # we need all backend indices, since we will iterate a specific order
+    # of dispatch keys looking for registrations.
+    backend_indices: Dict[DispatchKey, BackendIndex]
+
+    # a graph of dependency relations between operators.
+    graph: CompositeGraph
+
+    @staticmethod
+    def new(
+        dispatch_key: DispatchKey,
+        backend_indices: Dict[DispatchKey, BackendIndex],
+        grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
+    ) -> "DispatchlessComposite":
+        graph = get_graph(grouped_native_functions, backend_indices[dispatch_key])
+        return DispatchlessComposite(dispatch_key, backend_indices, graph)
+
+    @property
+    def backend_index(self) -> BackendIndex:
+        return self.backend_indices[self.dispatch_key]
+
+    # figures out what namespace to dispatch.
+    def _dispatch_namespace(
+        self, f: NativeFunction, g: Optional[NativeFunctionsGroup]
+    ) -> Optional[DispatchKey]:
+        # the precedence order is:
+        #   1. the current dispatch key we are generating code for
+        #   2. the CompositeExplicitAutograd kernel
+        #   3. the CompositeImplicitAutograd kernel
+        for k in (
+            self.dispatch_key,
+            DispatchKey.CompositeExplicitAutograd,
+            DispatchKey.CompositeImplicitAutograd,
+        ):
+            if has_registered_kernel(self.backend_indices[k], f, g):
+                return k
+
+        # bail if we have found no registered kernel.
+        return None
+
+    def _header_set(
+        self,
+        header_from_fn: Callable[[NativeFunction, Optional[NativeFunctionsGroup]], str],
+    ) -> List[str]:
+        return list(
+            {
+                header_from_fn(dep_f, dep_g)
+                for _, info in self.graph.items()
+                for dep_f, dep_g in info.dependencies
+            }
+        )
+
+    def aggregated_headers(self) -> List[str]:
+        def header_from(
+            dep_f: NativeFunction, dep_g: Optional[NativeFunctionsGroup]
+        ) -> str:
+            ns = self._dispatch_namespace(dep_f, dep_g)
+            prefix = f"{ns}" if ns is not None else ""
+            return f"#include <ATen/{prefix}Functions.h>"
+
+        return self._header_set(header_from)
+
+    def operator_headers(self) -> List[str]:
+        def header_from(
+            dep_f: NativeFunction, dep_g: Optional[NativeFunctionsGroup]
+        ) -> str:
+            ns = self._dispatch_namespace(dep_f, dep_g)
+            suffix = f"_{ns.lower()}_dispatch" if ns is not None else ""
+            return f"#include <ATen/ops/{dep_f.root_name}{suffix}.h>"
+
+        return self._header_set(header_from)
```

Comment:
You should also include the `ATen/ops/{root_name}_native.h` header for the operator being implemented.

Reply:
Ah, you're right! Thanks.